File: day02/Merge_Sort.py
import random
import time

def merge_sort(sort_list):
    if len(sort_list) == 1:
        return sort_list
    idx = len(sort_list)//2
    pre_list = merge_sort(sort_list[:idx])
    tail_list = merge_sort(sort_list[idx:])
    return sort(pre_list, tail_list)


def sort(pre_list, tail_list):
    sort_list = list()
    pre_len = len(pre_list)
    tail_len = len(tail_list)
    pre_idx = 0
    tail_idx = 0
    for i in range(0, pre_len + tail_len):
        if pre_idx < pre_len and tail_idx < tail_len:
            if pre_list[pre_idx] <= tail_list[tail_idx]:
                sort_list.append(pre_list[pre_idx])
                pre_idx +=1
            else:
                sort_list.append(tail_list[tail_idx])
                tail_idx +=1
        elif pre_idx == pre_len:
            sort_list.append(tail_list[tail_idx])
            tail_idx +=1
        elif tail_idx == tail_len:
            sort_list.append(pre_list[pre_idx])
            pre_idx +=1
    return sort_list

if __name__ == "__main__":
    l = list()
    for i in range(0,10000000):
        l.append(random.randint(0,3000000))
    #print(l)
    start = time.time()
    sort_l = merge_sort(l)
    end = time.time()
    print("sort time is %s seconds"%(end-start))
    #print(sort_l)

File: day02/Merge_Sort.c
#include "Merge_Sort.h"

int* merge_sort(int *sort_array, unsigned int len){
    int *pre_array, *tail_array;
    int pre_len,tail_len;
    int pre_idx,tail_idx;
    int i;
    int *p_array = NULL;
    if(len == 0||sort_array == NULL){
        return NULL;
    }
    if(len == 1){
        return sort_array;
    }
    pre_len = len/2;
    tail_len = len/2+len%2;
    pre_array = merge_sort(sort_array, pre_len);
    tail_array = merge_sort(sort_array+pre_len, tail_len);

    p_array = (int*)malloc(sizeof(int)*(len));
    if(p_array == NULL){
        return NULL;
    }
    pre_idx = 0;
    tail_idx = 0;
    for(i = 0;i< len;++i){
        if(pre_idx<pre_len&&tail_idx<tail_len){
            if(pre_array[pre_idx]<=tail_array[tail_idx]){
                p_array[i] = pre_array[pre_idx];
                pre_idx++;
            }else{
                p_array[i] = tail_array[tail_idx];
                tail_idx++;
            }
        }else if(pre_idx == pre_len){
            p_array[i] = tail_array[tail_idx];
            tail_idx++;
        }else if(tail_idx == tail_len){
            p_array[i] = pre_array[pre_idx];
            pre_idx++;
        }
    }
    memcpy(sort_array, p_array, len*sizeof(int));
    free(p_array);
    return sort_array;
}

